### ISSUE TITLE
Populate AttributeRouteInfo Name

### ISSUE THREAD

**Issue by RandyBuchholz** [aspnet/Mvc#5290]
## Issue
AttributeRouteInfo is not populated.

var routeName = ControllerContext.ActionDescriptor.AttributeRouteInfo.Name; is always null

## Discussion
AttributeRouting using Name is very powerful. The current implementation of AttributRouteingInfo holds back some important capabilities. It seems that AttributeRoute is somewhat second class right now because to get to it you need to use to MethodInfo.CustomAttributes to get to it. I don't think that RouteAttributes are really "Custom" anymore. It looks like that is improving with the AttributeRouteInfo class, but I'm not seeing that this class is getting populated. I would like to raise the priority of work in this area.

## Key Use Case
One of the things I see come up often is a desire to place views in different locations in the project structure. With core this has become much easier using ViewLocationExpanders. One of the key drivers for this in Domain Driven approaches. The Name attribute gives us a neat way to move towards domain (or other, like service-based) approaches to routing and project structure. For example, I can give a route a "pseudo-namespace" or a service-based perspective by doing something like Name="Domain.Provider.Service.Operation". This approach also helps me separate my business perspective from implementation details - I can base my design around hierarchal route structures aligned with the business/user view. This frees the controller developers to use a more technical/implementation approach to controller and action naming, providing an additional layer of abstraction to my system.

Within a ViewLocationExpander I can use this Name to build view paths. For example, I can do: 
var routeName = context.ActionContext.ActionDescriptor.AttributeRouteInfo.Name; 
var segments = routeName.Split('.');
var DomainProviderViewPath = $"{segments[0]}/{segments[1]}/Views";
Now it is easy to place my views in domain folders and for the application to find them.

## Strategy
I think you can see how using dot conventions in Name opens up a lot of possibilities for advanced routing and structuring. I can even see built-in capabilities for the framework to understand dots, and I am working on some of these concepts. I think there is a lot of discussion that would need to take place for this though. In the short term it would be nice to just get the AttributeRouteInfo populated so we don't have to parse this ourselves all of the time.

BTW - I would like to contribute to making this happen, but I haven't done any work on the project before. If someone could shoot me a message and help me get started that would be great.

**Comment by RandyBuchholz** [comment]
I've tried to use it in two cases where I thought it should be straightforward. The first is just inside a basic action:

```
    public class HomeController : Controller
    {
		[Route("~/")]
		[Route("NamedRoute", Name ="RouteName")]
		public IActionResult Index()
        {
			var name = ControllerContext.ActionDescriptor.AttributeRouteInfo.Name;
            return View();
        }
    }

```
When I step through, for AttributeRouteInfo the values are Name=null, Order=0, Template ="".

I also try to read AttributeRouteInfo in a ViewLocationExpander off ViewLocationExpanderContext :
```	
public class CustomViewLocationExpander : IViewLocationExpander {
		public virtual IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context, IEnumerable<string> viewLocations) {
			var descriptor = context.ActionContext.ActionDescriptor as ControllerActionDescriptor;
			var name = descriptor.AttributeRouteInfo.Name;
			var additionalLocations = BuildViewPaths(name);
			return additionalLocations;
		}
	}
```
I get the null/empty values again. 

To get the name of the route I need to do something like this:

```
	public class CustomViewLocationExpander : IViewLocationExpander {
		public virtual IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context, IEnumerable<string> viewLocations) {
			string routeName = string.Empty;
			var descriptor = context.ActionContext.ActionDescriptor as ControllerActionDescriptor;
			var routeNameFromMethodInfo = descriptor.MethodInfo.CustomAttributes;
			foreach (var routeAttribute in routeNameFromMethodInfo) {
				var one = routeAttribute.NamedArguments.FirstOrDefault(name => name.MemberName == "Name").TypedValue;
				if (one.Value != null) routeName = one.Value.ToString();
			}
			var additionalLocations = BuildViewPaths(routeName);
			return additionalLocations;
		}
	}

```
(Is there a more direct way to get this?)

It may be I'm misunderstanding the purpose, but from the name I would expect that anytime ActionContext is not null and I have an AttributeRoute on the action I should be able to get this value.

**Comment by dougbu** [comment]
Looks like the problem is here:
``` c#
[Route("~/")]
[Route("NamedRoute", Name ="RouteName")]
```

You've got two separate routes and are testing with a request that matches the first (`[Route("~/")]`). If the template you want is `"~/"`, use `[Route("~/", Name = "RouteName")]`.

BTW your "workaround" only works because you're examining all `RouteAttribute`s on the action, not just the one which matched the request. And an absolute route like `"~/"` will combine to match `string.Empty`.

**Comment by RandyBuchholz** [comment]
Thanks Doug, my bad. It's funny, there are two things I was doing here I've been doing for so long, I forgot they were so wrong.

I've used [Route("~/")] as "special case" during dev that sort of meant "Debugger Start Action". I used it like you use the "Project>Properties>Web>StartUrl" setting in a traditional asp.net app. Over time I guess I just quit thinking of it as an actual route that got matched and more like a debugger/ide command. 

The other wrong thing in my head was the idea that only one of the AttributeRoutes could have the Name property set. I have design patterns that use 1..1 relationships between a route name and an action, but use multiple templates on the action. I guess I over time I forgot that was just my design rule, not a framework rule. Though clearly wrong, I would have probably even said you would get an error when the route table was being built if you had two route attributes with Name="..." . LOL

So I guess subconsciously I saw the ~/ not as a routing but the debugger "magically" taking me to the Home/Index action, that could only have one Name= property, (no matter how many RouteAttributes), and AttributeRouteInfo.Name would give that to me. Strange how the way we (mis)use something can change the way we think about how it actually works.
